📚 Update docs on writing docs

ambirdsall opened the pull request:
## What
We have several custom extensions to basic markdown syntax to allow for nice features and environment-appropriate rendering differences between the full documentation site and in-app rendering of connector docs; but we don't really document how to use them anywhere. This changes that! It also closes https://github.com/airbytehq/airbyte/issues/31436.

In doing so, I also noticed that a link to the community code of conduct which was accidentally deleted when the file was moved as part of a doc reorganization; it's great that the old link was removed along with the file move that would have broken it, but since its replacement was never put in its place, there was one sentence that just sort of cut off mid-


## 🚨 User Impact 🚨
Hopefully, better documentation 😄 

## Pre-merge Actions
N/A

lmossman commented on docs/contributing-to-airbyte/writing-docs.md:
````diff
@@ -64,7 +64,7 @@ To make complex changes or edit multiple files, edit the files on your local mac
    ```
 
-   Then navigate to [http://localhost:3000/](http://localhost:3000/). Whenever you make and save changes, you will see them reflected in the server. You can stop the running server in OSX/Linux by pressing `Ctrl-C` in the terminal.  
+   Then navigate to [http://localhost:3005/](http://localhost:3005/). Whenever you make and save changes, you will see them reflected in the server. You can stop the running server in OSX/Linux by pressing `Ctrl-C` in the terminal.  
````
Should the `localhost:3000` below on line 76 also be changed to `localhost:3005`?

ambirdsall replied:
Nope! I had the same thought, and actually had to go back and fix my initial rewrite after double-checking the docs directory's `package.json`: we pass a `--port 3005` flag to the dev server, but the prod-style one uses docusaurus's default of `3000`.

lmossman commented on docs/contributing-to-airbyte/writing-docs.md:
```diff
@@ -84,6 +84,72 @@ To make complex changes or edit multiple files, edit the files on your local mac
+https://docs.airbyte.com may bury key details when rendered as inline documentation within
+the Airbyte application. In order to support all these different contexts
+without resorting to multiple overlapping files that must be maintained in parallel,
+Airbyte's documentation tooling supports multiple nonstandard features. Please familiarize
```
```suggestion
Airbyte's documentation tooling supports multiple nonstandard features.

Please familiarize
```
Add a line break here to help improve readability

lmossman replied:
Also, is there a specific reason why you put each line of text on a separate line of markdown here? If not, we should just using single markdown lines for each paragraph to match the approach used in the rest of the file, and to make editing easier.

ambirdsall replied:
> is there a specific reason why you put each line of text on a separate line of markdown here?

Yes, but not one that justifies leaving it that way—it's easier to navigate multiple lines than a single wrapped line in old-fashioned neckbeard text editors, and the resulting html is identical. Will update, no problem.

lmossman commented on docs/contributing-to-airbyte/writing-docs.md:
```diff
@@ -84,6 +84,72 @@ To make complex changes or edit multiple files, edit the files on your local mac
+but have no impact on https://docs.airbyte.com.
+
+#### Jump to the relevant documentation section when specific Connector Builder inputs are focused with `<FieldAnchor>`
+In the documentation, the relevant section needs to be wrapped in a `<FieldAnchor field="path.to.field" />` component. These are rendered as regular divs in the documentation site, so they have no effect in places other than the in-app documentation panel—however, note that there must be blank lines between a custom tag like `FieldAnchor` the content it wraps for the documentation site to render  ...
```
Could we add a note here indicating that it is the field path underneath connectionSpecification, to make this very clear? Otherwise reader's may be unsure if they should include that in the path

lmossman reviewed (approved):
Thanks for putting this change up, this will definitely be useful for connector developers!

Added a few suggestions but nothing really blocking so feel free to merge after they are addressed, or I'm happy to take another look if you want

ambirdsall commented on docs/contributing-to-airbyte/writing-docs.md:
```diff
@@ -84,6 +84,152 @@ To make complex changes or edit multiple files, edit the files on your local mac
+Please familiarize yourself with all the tools available to you when writing documentation for a connector, so that you can provide appropriately tailored information to your readers in whichever context they see it. 
+
+:::note
+As a general rule, features that introduce new behavior or prevent certain content from rendering will affect how the Airbyte UI displays markdown content, but have no impact on https://docs.airbyte.com. If you want to test out these in-app features in a local Airbyte build, ensure that you have the `airbyte` git repository checked out to the same parent directory as the airbyte platform reposito ...
```
New note about local development added here. I thought about making it a separate `:::tip`, but that seemed like it might look a bit too busy.

lmossman replied:
To make it very explicit how to actually do this, we should add a link to [this page](https://docs.airbyte.com/contributing-to-airbyte/resources/developing-locally/#develop-on-airbyte-webapp) in this note. Should be safe to do since that is a public page

lmossman commented on docs/contributing-to-airbyte/writing-docs.md:
```diff
@@ -84,6 +84,152 @@ To make complex changes or edit multiple files, edit the files on your local mac
+:::
+
+#### Jump to the relevant documentation section when specific Connector Builder inputs are focused with `<FieldAnchor>`
+In the documentation, the relevant section needs to be wrapped in a `<FieldAnchor field="path.to.field" />` component. These are rendered as regular divs in the documentation site, so they have no effect in places other than the in-app documentation panel—however, note that there must be blank lines between a custom tag like `FieldAnchor` the content it wraps for the documentation site to render  ...
```
Could we add a note here that the entire piece of documentation that is relevant to the field should be wrapped, _not_ just the header of the relevant section, since the wrapping defines what will be highlighted

lmossman commented on docs/contributing-to-airbyte/writing-docs.md:
```diff
@@ -84,6 +84,152 @@ To make complex changes or edit multiple files, edit the files on your local mac
+#### Jump to the relevant documentation section when specific Connector Builder inputs are focused with `<FieldAnchor>`
+In the documentation, the relevant section needs to be wrapped in a `<FieldAnchor field="path.to.field" />` component. These are rendered as regular divs in the documentation site, so they have no effect in places other than the in-app documentation panel—however, note that there must be blank lines between a custom tag like `FieldAnchor` the content it wraps for the documentation site to render  ...
+
+The `field` attribute must be a valid json path to one of the properties nested under `connectionSpecification.properties` in that connector's `spec.json` or `spec.yaml` file. To mark a section as highlighted after the user picks an option from a `oneOf`: use a `field` prop like `path.to.field[value-of-selection-key]`. It's also possible to highlight the same section for multiple fields by separa ...
```
Could we explicitly call out that the `connectionSpecification.properties` part should _not_ be included in the `field` path? I think a concrete example here would help make this super clear

lmossman replied:
For the `oneOf` callout, let's add a bit more explanation: `value-of-selection-key` isn't really clear to non-developers, since "selection key" is really just a term used in the code.

Instead we should call out that what needs to go in the brackets is the corresponding `const` value of the option being targeted. Again, a concrete example here would be very helpful

lmossman commented on docs/contributing-to-airbyte/writing-docs.md:
````diff
@@ -84,6 +84,152 @@ To make complex changes or edit multiple files, edit the files on your local mac
+:::
+
+```md
+#### Collapsible content with `<details>` and `<summary>`
````
nit: add a newline here

lmossman reviewed (commented):
Had some more suggestions, getting close though!

ambirdsall commented:
@lmossman thanks for showing this to other other other Alex, that feedback was really helpful. I addressed all the comments you left, and cleaned up one accidentally-duplicated example besides; hopefully it's now good to go.

lmossman commented on docs/contributing-to-airbyte/writing-docs.md:
```diff
@@ -84,6 +84,146 @@ To make complex changes or edit multiple files, edit the files on your local mac
+#### Jump to the relevant documentation section when specific Connector Builder inputs are focused with `<FieldAnchor>`
+In the documentation, the relevant section needs to be wrapped in a `<FieldAnchor field="path.to.field" />` component. When a user focuses the field identified by the `field` attribute in the Connector Builder UI, the documentation pane will automatically scroll to the associated section of the documentation, highlighting all content contained inside the `<FieldAnchor></FieldAnchor>` tag. These a ...
+
+The `field` attribute must be a valid json path to one of the properties nested under `connectionSpecification.properties` in that connector's `spec.json` or `spec.yaml` file. For example, if the connector spec contains a `connectionSpecification.properties.replication_method.replication_slot`, you would mark the start of the related documentation section with `<FieldAnchor field="replication_met ...
```
```suggestion
The `field` attribute must be a valid json path to one of the properties nested under `connectionSpecification.properties` in that connector's `spec.json` or `spec.yaml` file. For example, if the connector spec contains a `connectionSpecification.properties.replication_method.replication_slot`, you would mark the start of the related documentation section with `<FieldAnchor field="replication_method.replication_slot" />` and its end with `</FieldAnchor>`. To mark a section as highlighted after the user picks an option from a `oneOf`: use a `field` prop like `path.to.field[value-of-selection-key]`, where the `value-of-selection-key` is the value of a `const` field nested inside that `oneOf`; for example, if the specification of the oneOf field is:
\```
"replication_method": {
  "type": "object",
  "title": "Update Method",
  "oneOf": [
    {
      "title": "Read Changes using Binary Log (CDC)",
      "required": ["method"],
      "properties": {
        "method": {
          "type": "string",
          "const": "CDC",
          "order": 0
        },
        "initial_waiting_seconds": {
          "type": "integer",
          "title": "Initial Waiting Time in Seconds (Advanced)",
        },
      }
    },
    {
      "title": "Scan Changes with User Defined Cursor",
      "required": ["method"],
      "properties": {
        "method": {
          "type": "string",
          "const": "STANDARD",
          "order": 0
        }
      }
    }
  ]
}
\```
then the const values are `CDC` and `STANDARD`, so  you can wrap a specific replication method's documentation section with a `<FieldAnchor field="replication_method[CDC]">...</FieldAnchor>` tag, and it will be highlighted if the user selects that replication method in the UI. It's also possible to highlight the same section for multiple fields by separating them with commas, like `<FieldAnchor path="replication_method.replication_slot,replication_method.queue_size">`.
```

I think an even more explicit example like this would be helpful. Lmk what you think

ambirdsall replied:
I do think it would be helpful, though I do wonder if it's _too_ much detail: if we ever change the structure of the connector specification, odds are very slim we'd remember to keep this snippet of documentation up-to-date. But I think I'm probably overthinking it there: the expanded example is clearly helpful now, so I'll add it in.

lmossman reviewed (approved):
@ambirdsall left a couple more small comments. Once those are addressed this looks good to merge 👍
